armbalancer.go

```diff
@@ -72,16 +72,43 @@ func New(opts Options) http.RoundTripper {
 	}
 
 	if opts.TransportFactory == nil {
-		opts.TransportFactory = newRecyclableTransport
+		opts.TransportFactory = make(map[string]func(id int, parent *http.Transport, host string, port string, recycleThreshold, minReqsBeforeRecycle int64) http.RoundTripper)
+		opts.TransportFactory[strings.ToLower(host+":"+port)] = newRecyclableTransport
 	}
 
-	t := &transportPool{pool: make([]http.RoundTripper, opts.PoolSize)}
-	for i := range t.pool {
-		t.pool[i] = newRecyclableTransport(i, opts.Transport, host, port, opts.RecycleThreshold, opts.MinReqsBeforeRecycle)
+	t := &pooledTransport{pool: make(map[string]*transportPool)}
+	for key, factory := range opts.TransportFactory {
+		transports := make([]http.RoundTripper, 0, opts.PoolSize)
+		for i := 0; i < opts.PoolSize; i++ {
+			transports = append(transports, factory(i, opts.Transport, host, port, opts.RecycleThreshold, opts.MinReqsBeforeRecycle))
+			t.pool[key] = &transportPool{pool: transports}
+		}
 	}
 	return t
 }
 
+type pooledTransport struct {
+	pool map[string]*transportPool
+}
+
+func (t *pooledTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	transportPool, err := t.compareHosts(req.URL)
+	if err != nil {
+		return nil, err
+	}
+	return transportPool.RoundTrip(req)
+}
+
+func (t *pooledTransport) compareHosts(req *url.URL) (*transportPool, error) {
+	parsedHostName := req.Hostname()
+	port := req.Port()
+	transportPool, ok := t.pool[strings.ToLower(parsedHostName+":"+port)]
+	if !ok {
+		return nil, fmt.Errorf("host %q is not supported by the configured ARM balancer, supported host name is %+v", req.Host, t.pool)
```

Comment:
what does the output of the message look like? Want to make sure we don't leak implementation details with `%+v`. I'm fine just dropping the `"supported host name is %+v"` part of the message if it's easier.